sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2012,6 +2012,152 @@ class Dataset[T] private[sql](
   @scala.annotation.varargs
   def agg(expr: Column, exprs: Column*): DataFrame = groupBy().agg(expr, exprs : _*)
 
+  /**
+   * (Scala-specific)
+   * Unpivot a DataFrame from wide format to long format, optionally
+   * leaving identifier variables set.
+   *
+   * This function is useful to massage a DataFrame into a format where some
+   * columns are identifier variables (`ids`), while all other columns,
+   * considered measured variables (`values`), are "unpivoted" to the rows,
+   * leaving just two non-identifier columns, 'variable' and 'value'.
+   *
+   * {{{
+   *   val df = Seq((1, 11, 12L), (2, 21, 22L)).toDF("id", "int", "long")
+   *   df.show()
+   *   // output:
+   *   // +---+---+----+
+   *   // | id|int|long|
+   *   // +---+---+----+
+   *   // |  1| 11|  12|
+   *   // |  2| 21|  22|
+   *   // +---+---+----+
+   *
+   *   df.melt(Seq("id")).show()
+   *   // output:
+   *   // +---+--------+-----+
+   *   // | id|variable|value|
```

Comment:
are `variable` and `value` from other DBMSes? it would be good to match the column names with them instead of coming up with our own naming.

Reply:
I could remove the default values for `variableColumnName` and `valueColumnName` from `Dataset.melt` as they are generally not useful (may be changed by users anyway). And in SQL they have to be given as well. This removes one overloaded `melt` from `Dataset`.

Reply:
Looks reasonable to me. If the SQL syntax requires to set these 2 names, Scala API should do so as well.

Reply:
done